doc/PROTOCOL-SEMANTICS.md

```diff
@@ -0,0 +1,268 @@
+passing, not object orientation.
+
+Related Methods are typically grouped into a Service. To gRPC, a Service is a
+group of methods that tend to be implemented together and that all share the
+Service's namespace. A Service is a higher-level abstraction and may not be
+present explicitly in all implementations. However, the namespace provided by a
+Service is a core distinguishing feature of its Methods; if two Methods have the
+same name but exist in different Services they must be considered distinct and
+not be confused. A Method name including its Service namespace prefix with a "/"
+separator is a "full method name".
+
+## Calls
+
+RPCs, or "Calls," are initiated by a client to a server, typically via a
+Channel. There may be multiple servers that _could_ have received the Call (as
+is common for load balancing), but only a single server may process an
+individual Call. Calls are assumed to be non-idempotent and may not be
+"replayed" except for when gRPC is explicitly informed it is safe to do so.
+
+Calls are natively two independent streams (i.e., full duplex bidirectional) of
+Messages. The request stream is started with Request Headers and ended by Half
+Close. The response stream is started with Response Headers and ended by
+Trailers, or consistes only of Trailers. Messages may exist between the headers
+and the end of the stream. Request Headers, Response Headers, Messages, Half
+Close, and Trailers are the units of communication and, absent the Call's
+termination, will be communicated to the remote without the need to send further
+units on the stream. However, see the optimizations permitted for unary Calls
+below.
+
+Request Headers contain the Full Method Name and Metadata. Response Headers
+contain Metadata. Trailers contain the Status and Metadata. Messages contain the
+Message Payload. These contents are not exhaustive; gRPC features may extend
+these concepts. It is quite common for features to add additional fields to
+Request Headers, Response Headers, and Trailers.
+
+The Call initiation is with Request Headers, within which the client
+indicates the method to be run by its Full Method Name. The Call is gracefully
+completed when the server responds with Trailers, which contains a Status
+communicating the success or failure of the RPC. If a server responds with
+Trailers before receiving the client's Half Close, then any unprocessed
+client-sent Messages and Half Close is lost. Note that on the server there is a
+period of time between when the server application responds with a Trailers and
+when that Trailers is actually sent; the Call is only truly complete when the
+Trailers is sent. Similarly, on the client there is a period between the gRPC
+implementation receiving the Trailers and when the application receives the
+Trailers; the Call is only truly complete when the Trailers is received by the
+application.
+
+Calls may terminate early by being "cancelled." Implementations must allow
```

Comment:
MUST?  MAY (later)?

I'd say be consistent and either all-caps all these words when they appear or lowercase the SHOULD above.

Reply:
The SHOULD above is the only one that looks to be uppercase. I didn't find any other must, may, or should that is upper case. They SHOULD be lowercase. :smile: